### DISCUSSION TITLE
Use pandas to develop a UI component library

### DISCUSSION THREAD

**Discussion by Darcrandex** [chakra-ui/panda#826] [Q&A] [answered]
Hello, when I create a component library based on pandas and package it for release, I find that the packaged file does not contain the classes used in the component. What should I do?

For example, my button component does not contain color style code after packaging.

```
<button
        className={css({
          color: "blue.600",
          padding: "2",
          bg: "blue.100",
          borderRadius: "md",
        })}
      >
        {props.children}
 </button>
```

[color: "blue.600"](https://github.com/Darcrandex/bug-ui/blob/96f4fd74df371d4a973e4f0ece12197e807fb29a/src/components/Button/index.tsx#L17)
 does not generate the corresponding css code

repository and npm
https://github.com/Darcrandex/bug-ui
https://www.npmjs.com/package/@darcrand/bug-ui

**Comment by segunadebayo** [comment, marked as answer]
Hi @Darcrandex, we wrote a guide here:
https://panda-css.com/docs/guides/component-library

**Reply by ghost** [reply]
this is EXTREMELY useful, but im struggling nevertheless to get the mono repo approach described to work. it would be a huge help if there was a demo repo somewhere! 

an example of the sort of dumb question that im hung up on - i'm consuming the library in a next.js app - so do i set up the library to use postcss or not?

**Reply by astahmer** [reply]
this monorepo example might help https://github.com/astahmer/panda-monorepo/

**Reply by ghost** [reply]
THANK YOU this is unbelievably useful. will study closely. BRILLIANT. so grateful.

**Reply by hainessss** [reply]
@segunadebayo  or @astahmer  could you maybe to speak to why in the monorepo approach we create a separate styled-system package and all the apps generate into that package? is this just if you want to share recipes created in one package with all the others?

**Reply by astahmer** [reply]
so that you don't bundle twice your `styled-system` folders. 
that can happen if in your app you import once from a local folder (`../styled-system`) and once from your monorepo package (`@acme/styled-system`), you'd end up with the `css` fn shipped twice because these are different files after all

for example if you have a preset with a bunch of design-system recipes but you need a custom one that is not part of the preset, you need a local `styled-system` folder for that since the one from the component library (`@acme/styled-system`) won't contain that specific recipe files

to solve that, you can also use bundler aliases like [resolve.alias](https://vitejs.dev/config/shared-options#resolve-alias) to point your monorepo package path (`@acme/styled-system`) to your local (`../styled-system`) folder and always import from the monorepo package path (`@acme/styled-system`)

**Reply by hainessss** [reply]
thanks for this explanation @astahmer !

**Comment by VinSpee** [comment]
Hi! I'm still struggling with this.

ideally, my _app_ (consumer of my panda package) shouldn't have to know about panda at all. How can I achieve this? I can't find a way which doesn't require a `panda.config.ts` and `global.css` in the `app` (consuming the panda package)

**Reply by astahmer** [reply]
does [this](https://stately.ai/registry/editor/d30cb4f3-2279-41a6-ad95-13e72ccfe884?machineId=0ff5a34b-1d63-476d-ad2d-21c91dad9973&mode=Simulate) help ?

preview here:
![studio-0ff5a34b-1d63-476d-ad2d-21c91dad9973-dark](https://github.com/chakra-ui/panda/assets/47224540/439e9720-c12a-4624-9571-e5f4e6c02238)

for your use-case, if your users won't use panda, just ship a static css along with your components and bundle your `styled-system`  package with it (= don't make it external)

**Reply by VinSpee** [reply]
@astahmer extremely helpful actually. Love the state chart approach to the explanation. Having the consuming libraries consume a shared preset was the piece I was missing. Thanks!

**Reply by coopbri** [reply]
I think I'm missing something simple here. My use case seems very similar to @VinSpee's -- a Panda component library completely separate from downstream app implementation (in my case at least, no monorepo setup), without requiring the apps to use Panda at all. I am using the [static CSS file method](https://panda-css.com/docs/guides/component-library#ship-a-static-css-file). This works, but the output CSS does not have classes for all of the base styles/CSS variables since Panda only builds the classes it needs to based on what CSS I use in the library.

For example, I have a `Button` component in my library that sets some styles in a recipe (padding 4 for example). I am able to use `p={4}` in the app code since the `p_4` exists from the build, but not, for example, `p={3}` since the `p_3` class isn't generated at build time (which makes sense). I did see the caveat in the docs about CSS customization too ("You can't customize the styles since the css is already generated"). I do think I understand the issue here, I'm just curious if there is a way to get this working without requiring the apps to use Panda as well. I also referenced the state chart mentioned above.

I assume I am missing something related to shipping the `styled-system` package. I tried working with the Panda `emitPackage` option, and have the following exported from my library:

```ts
export * from "<OUTDIR_PATH>/css";
export * from "<OUTDIR_PATH>/jsx";
export * from "<OUTDIR_PATH>/patterns";
export * from "<OUTDIR_PATH>/tokens";
```

**EDIT:** there is a relevant discussion in the Panda CSS Discord server [here](https://discord.com/channels/1118988919804010566/1118988920533823540/1119936246114943046). For now, I am thinking of just using Panda in downstream applications (it's essentially just a thin config & compilation layer at that point anyways), or statically generating all possible atomic utility classes for the Panda builtin utilities (e.g. every atomic padding class) via the `staticCss` config option. With that being said, I would love to avoid either of these solutions if it's possible to generate atomic classes for every possible case, so that I can change any JSX styles in the downstream apps on-the-fly (although that may be an antipattern).

**EDIT 2:** I ended up just creating a Panda preset in my component library code (as with the solution mentioned above in this thread), installing Panda in downstream applications, and injecting the preset in those app Panda configs. This way I can avoid shipping a giant CSS bundle, keeping the benefits of truly atomic, JIT-generated CSS. Small price to pay at the end of the day :panda_face:

**Reply by willTastyMakers** [reply]
@coopbri,  Hi mate, I too am trying to build a stand alone component library using  PandCSS. I am using Pandacss and ark-ui/React. I've created some custom react components as well as ark-ui  components with pandacss's recipe.  My distributable  npm package contains my jsx components, recipes, and my preset and such.  My intention is to have all of my downstream apps include panda npm packages use the library's panda preset and of course actual jsx components.  Everything packages correct;y, but I am having issue  consuming the npm in the downstream.  

[ERROR] Could not resolve "@tms/design-system/styled-system/jsx" and" @tms/design-system/styled-system/recipe".  In the libraries, tsup build process I have adde the  "--external ***** '  configs to the package.json build scripts and  I have tried placing them in the  tsup.config file as well, but all to no avail.  

When you said you are generating Static CSS in the component library, does that mean you are sharing the Static CSS and the Components?

**Reply by astahmer** [reply]
(pasting from discord so it can be found here as well)

if you intend for your users to also use panda:
-> publish a preset (since you're using a custom theme)
-> set  emitPackage: true in your library's panda.config, and a custom outdir
-> set that outdir as --external so it won't be part of your dist (since you want your users to re-generate it from their side, keeping the benefits JIT-generated CSS)
-> publish the rest of your DS components as usual (using tsup & a dist folder would be fine)
-> tell your users to import your panda preset, use emitPackage: true and the same outdir as you, then tell them to use panda codegen command

otherwise, if they will not use panda:
-> publish a css file generated from the panda cssgen --outfile dist/styles.css command
-> publish your DS components as usual, you do NOT need emitPackage: true and do NOT need a outdir, quite the contrary you need your outdir to be bundled into your dist (using tsup or equivalent)

**Reply by willTastyMakers** [reply]
@alex thank you for responding....  I believe I have followed the scenario you have described under  **'yes'**

### **// my component library package.json**
```

script {
...
"build": "tsup --format cjs,esm --dts  --external @tms/styled-system  && pnpm ship",
...
}
{
  "name": "@tms/design-system",
  "files": ["src", "dist"]
}
```

### **//my component  library  panda.config.ts**

```
import { defineConfig } from '@pandacss/dev';
import {tmsPandaPreset} from './src/theme/tms-panda-preset';

export default defineConfig({
  preflight: false,
  presets: ['@pandacss/dev/presets', tmsPandaPreset],
  include: [
         './src/**/*.{js,jsx,ts,tsx}', 
         './stories/**/*.{js,jsx,ts,tsx}'
   ],
  exclude: [],
  jsxFramework: 'react',
  emitPackage: true,
  outdir: '@tms/styled-system'
});
```

### //my tsup.config.ts

```
  external: [
      '@pandacss/dev',
      '@tms/styled-system'
  ],
  format: ['cjs', 'esm'],
  entry: ['src/index.ts'],
  dts: true,
  sourcemap: true,
  clean: true,
```


### **The consuming next.js app using pandacss**
```  
import { defineConfig } from '@pandacss/dev';
import pandaPreset from '@pandacss/preset-panda'
import {tmsPandaPreset} from '@tms/design-system';

preflight: true,
    presets: [pandaPreset, tmsPandaPreset],
    include: [
        './node_modules/@tms/design-system/src/**/*.{js,jsx,ts,tsx}',
        './app/**/*.{js,jsx,ts,tsx}',
        './component/**/*.{js,jsx,ts,tsx}',
        './lib/**/*.{js,jsx,ts,tsx}'
    ],
    exclude: [],
    jsxFramework: 'react',
    emitPackage: true,
    outdir: '@tms/styled-system',
```
  
When ever I try to run `pnpm install` in the consuming app I get the following error
```
🏃 [prepare] Running command "panda codegen && husky install"
✘ [ERROR] Could not resolve "@tms/styled-system/jsx"

    node_modules/.pnpm/@tms+design-system@1.0.5_@babel+core@7.22.9_@internationalized+date@3.3.0_react-dom@18.2.0_react@18.2.0/node_modules/@tmsdesign-system/dist/index.mjs:4:7:
      4 │ } from "@tms/styled-system/jsx";
        ╵        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "@tms/styled-system/jsx" as external to exclude it from the bundle, which will remove this error.

✘ [ERROR] Could not resolve "@tasty-maker-studio/styled-system/recipes"

    node_modules/.pnpm/@tms+design-system@1.0.5_@babel+core@7.22.9_@internationalized+date@3.3.0_react-dom@18.2.0_react@18.2.0/node_modules/@tms/design-system/dist/index.mjs:7:7:
      7 │ } from "@tmsstyled-system/recipes";
        ╵        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "@tasty-maker-studio/styled-system/recipes" as external to exclude it from the bundle, which will remove this error.
```

I tried to follow  the  Docs  [Using Panda as External Package](https://panda-css.com/docs/guides/component-library#use-panda-as-external-package)..

I have tried using --external in both the "script" of package.json and the tsup external[]...  is there something I am missing

**Reply by astahmer** [reply]
is it because you have a `prepare` script in your next app that does a `codegen` that prevents you from installing ? can you try running `pnpm i --ignore-scripts` first, then run `panda codegen` and finally just start your app ?

if that does not work, please provide a minimal repro

**Reply by willTastyMakers** [reply]
@astahmer  thanks for sticking with me on this one... 
I tried to do as you suggested:
1. removed the  `prepare`  script
2. created another script called `panda-prep: pnpm panda codegen`

3.
```
->  pnpm i --ignore-scripts 
```
creates a successful install  of the package.json

4.
```
-> pnpm panda-prep
```

returns the same error 
```
🏃 [prepare] Running command "panda codegen && husky install"
✘ [ERROR] Could not resolve "@tms/styled-system/jsx"

    node_modules/.pnpm/@tms+design-system@1.0.5_@babel+core@7.22.9_@internationalized+date@3.3.0_react-dom@18.2.0_react@18.2.0/node_modules/@tmsdesign-system/dist/index.mjs:4:7:
      4 │ } from "@tms/styled-system/jsx";
        ╵        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "@tms/styled-system/jsx" as external to exclude it from the bundle, which will remove this error.

✘ [ERROR] Could not resolve "@tasty-maker-studio/styled-system/recipes"

    node_modules/.pnpm/@tms+design-system@1.0.5_@babel+core@7.22.9_@internationalized+date@3.3.0_react-dom@18.2.0_react@18.2.0/node_modules/@tms/design-system/dist/index.mjs:7:7:
      7 │ } from "@tmsstyled-system/recipes";
        ╵        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "@tasty-maker-studio/styled-system/recipes" as external to exclude it from the bundle, which will remove this error.
  ```

I will create 2 minimal repos 1 for the  **ui-library** and 1 for the **next app**
  
  Once again...  👏🏽 big  ups for helping me work through this...

**Reply by willTastyMakers** [reply]
@astahmer 
Sorry for the late response. When I  did it with a smaller next.js app I felt as though I was getting  a bit closer, but no such luck ... 🙁 ... so I've returned. 

 I created a  very minimal [repo called panda-tms](https://github.com/Tasty-Maker-Studio/panda-tms).  I can not provide the actual repo to the component library `@tasty-maker-studio/design-system` because it private and associated with github organization CI/CD.   I included a copy of the [@tasty-maker-studio'.](https://github.com/Tasty-Maker-Studio/panda-tms/tree/main/%40tasy-maker-studio)  as it would appear in node_modules.

I  can "sort of" get it to work if I :
 
1st:  begin with the  [panda.config.ts ](https://github.com/Tasty-Maker-Studio/panda-tms/blob/f90510dd961090ad33ce7bcf887ff8c2e4eb8e1b/panda.config.ts#L21) preset with the generic configuration of
```
presets: ['@pandacss/dev/presets'],
```
2nd: do the pnpm install and then run the panda prepare t
```
> pnpm install 
> pnpm panda-prep
```
The `panda-prep` script command runs  `panda codegen` injecting a styled-system in the node_modules `@tasty-maker-studio/styled-system` 

3rd:  return to panda.config.ts. and  comment out the preset of generic and  an uncomment the preset that includes both generic and the `tmsPandaPreset` 
```
    ...
    //presets: ['@pandacss/dev/presets'],
     presets:['@pandacss/dev/presets',  tmsPandaPreset],
     ...
```     

4th:  run  the panda prepare script again
```
-> pnpm panda-prep
```

Now the  `node_modules/@tasty-maker-studio/design-system`  will also have a[  @tasty-maker-studio@styled-component with correct recipes](https://github.com/Tasty-Maker-Studio/panda-tms/tree/main/%40tasy-maker-studio/design-system/src/theme/recipes). and will load **some ** of the components correctly in page.tsx.  The Dialog Drawers are not working as they do in the design-system's storybook.   I don't believe the way I am loading  `tmsPandaPreset` is correct?  I say that because :
- The recipes have to be imported directly  and added to the panda.config.ts even though they are import through the [ tmsPandaPreset](https://github.com/Tasty-Maker-Studio/panda-tms/blob/f90510dd961090ad33ce7bcf887ff8c2e4eb8e1b/%40tasy-maker-studio/design-system/src/theme/tms-panda-preset.ts#L4)
-  the [ keyFrames](https://github.com/Tasty-Maker-Studio/panda-tms/blob/f90510dd961090ad33ce7bcf887ff8c2e4eb8e1b/%40tasy-maker-studio/design-system/src/theme/tms-panda-preset.ts#L18) in the tmsPandaPreset do not seem to be present/working.


Thank you in advance for any direction or help you can provide 🙌🏽

**Reply by astahmer** [reply]
fixed it, please make smaller (and easier to install..) reproductions next time 🙏 
https://github.com/Tasty-Maker-Studio/panda-tms/pull/1/commits/476a02adfc4904c01fb4e01eaa8dcd8fd0e9208f

summary of the changes:

in DS package:

- added tsup entrypoint
- added panda.config + tsconfig
- used direct import for the preset instead of importing the index which itself import everything (other components etc)

in app (next):
- used direct import for the preset instead of importing the index which itself import everything (other components etc)
- changed the tsconfig.target for tsup to work properly

**Reply by willTastyMakers** [reply]
@astahmer ,
Thank you so very much for helping me with this issue.  I was able to apply your changes to my library and my consuming Nextjs application and successfully  install  and generate 'codegen'.  I apologize for the disorganization of the code. I was in the middle of assembling a minimal repo of a library ui and a consumer web app that was a little more straight forward , but I figured I had better present something quickly before my inquiries were lost in the shuffle. Next time I will be more diligent in provide better organized example 🙏

While the install assembly is working, I have noticed that the some of the  classes generated in the library's recipe variants and  a set of Keyframe animations in the  library's panda preset  are not present in the Next.js consumer app.  When I open the developer tools, I can see the classes on the html elements

**class="lateralDrawer lateralDrawer--variant_left lateralDrawer--size_md"**

```<div data-scope="dialog" data-part="container" id="dialog::Rdmcq::container" data-state="open" class="lateralDrawer lateralDrawer--variant_left lateralDrawer--size_md" style="">```

but when these classes are not present in the developer tool styles.  However, using the Dev Tools, the classes associated with the elements and their respective recipe css values are present in the component library when viewed through the storybook  "**.stories.tsx".

The only change I made from your code was in the library ( design-system) panda.config.ts.  There was an [import](https://github.com/astahmer/panda-tms/blob/476a02adfc4904c01fb4e01eaa8dcd8fd0e9208f/%40tasy-maker-studio/design-system/panda.config.ts#L4) 

```
// you need to separate the preset from the components otherwise esbuild & panda get confused
import { tmsPandaPreset } from "@tasty-maker-studio/design-system/src/theme/tms-panda-preset";
```

Perhaps stemming from the confusing organization of my code, I figured you were actually suggesting that the panda.config.ts in the 
component library ( "@tasty-maker-studio/design-system/...) should be importing the tmsPandaPreset like:

```
import { tmsPandaPreset } from './src/theme/tms-panda-preset';
```
Since the `@tasty-maker-studio/design-system` is produced by the component library and thus not available.  I only bring this up to make sure I did not miss something  and correctly  implementing your comment.

> you need to separate the preset from the components otherwise esbuild & panda get confused

Thanks 🙌🏽 for your help.

**Reply by astahmer** [reply]
> but when these classes are not present in the developer tool styles. 

that's weird, I have them when running `pnpm dev` in the repo root, which is a next app

![Screenshot 2023-08-02 at 21 50 34](https://github.com/chakra-ui/panda/assets/47224540/2759c585-8cf4-40b1-9d55-6d3e9bb624ad)

EDIT: nvm, got it, I can see some are missing

> import { tmsPandaPreset } from './src/theme/tms-panda-preset';

yes this is correct ! my bad there

**Reply by astahmer** [reply]
the issue is that you're importing from a re-export, not directly from the styled-system, so it's not being picked up as a recipe by the static extractor (we don't traverse files)

https://github.com/Tasty-Maker-Studio/panda-tms/blob/a4dd104f6430e2be22d4cddf4175c588b88c90c9/%40tasy-maker-studio/design-system/src/components/Overlay/Drawer/LateralDrawer.tsx#L16
https://github.com/Tasty-Maker-Studio/panda-tms/blob/a4dd104f6430e2be22d4cddf4175c588b88c90c9/%40tasy-maker-studio/design-system/src/components/Overlay/Drawer/DialogParts.ts#L35

if you remove the `lateralDrawerStyles` import from `DialogParts` in `LateralDrawer.tsx`, and instead import it directly like `import { lateralDrawer as lateralDrawerStyles } from "@tasty-maker-studio/styled-system/recipes";` then it will work as expected

**Reply by willTastyMakers** [reply]
Just saw your response I will give a go


@astahmer ,
When I open the lateral-drawer in the  design-system library and I view the dev tools I can see the class value generated by the recipe
```
export const lateralDrawer = defineRecipe({
  name: 'lateralDrawer',
  description: 'A lateral drawl  style',
  base: parts({
     .....
    content: {
      position: 'absolute',
      zIndex: '99999',
      background: 'neutral.50',
      height: '100%',
      minW: 'xs',
      overflowY: 'auto',
      border: 'none',
      width: '100%'
    },
...
  }),
  variants: {
    variant: {
   ..
      }),
      left: parts({
        content: {
          top: '0',
          bottom: '0',
          left: '0',
          '&[data-state=open]': { animation: 'slideRightIn 0.5s ease-out' },
          '&[data-state=closed]': { animation: 'slideLeftOut 0.5s ease-in' }
        },
        closeTrigger: {
          right: '0'
        }
      })
    },
    size: {
...
      md: parts({
        content: {
          width: '50%',
          minWidth: '400px'
        }
      }),
      ...
    }
  },
 ..
});
```
![PANDA-DEBUB](https://github.com/chakra-ui/panda/assets/58872627/31eec94a-2943-43ef-a091-75462f21da31)

**Reply by willTastyMakers** [reply]
@astahmer ,

I tried as your suggest and removing 

> the lateralDrawerStyles import from DialogParts in LateralDrawer.tsx, and instead import it directly like import { lateralDrawer as lateralDrawerStyles } from "@tasty-maker-studio/styled-system/recipes";

and importing from recipes directly
[DialogParts.tsx ](https://github.com/Tasty-Maker-Studio/panda-tms/blob/main/%40tasy-maker-studio/design-system/src/components/Overlay/Drawer/DialogParts.ts)
[LateralDrawer.tsx](https://github.com/Tasty-Maker-Studio/panda-tms/blob/main/%40tasy-maker-studio/design-system/src/components/Overlay/Drawer/LateralDrawer.tsx#L4)

but did not work. It still does not contain the classes that are displayed in the  screenshot above of the storybook. 
I will continue to hack away at it.

**Comment by willTastyMakers** [comment]
@astahmer ,
:pray: :man_dancing:
Thank you for all you help. I was able to solve my issues and now our team is back on track. I appreciate you... 
:raised_hands:

**Reply by astahmer** [reply]
awesome ! what was the issue ?

**Reply by willTastyMakers** [reply]
**2 issues.**
1. I did not realize there was a breaking change regrading recipes in v0.9.0 and when I switched from  'name' to 'className' that helped me sort a problem I was having in my github action workflow build.

2. Since I did not want matriculate the release version in the design-system package, I was building in design-system client repo and copy files from dist to the panda-tms/node_modules@tasty-maker-studio/design-system/dist. When I did this copy,  I forgot to also bring over the changes from  design-system/src as well.  

The changes were a part of your instructions
> if you remove the lateralDrawerStyles import from DialogParts in LateralDrawer.tsx, and instead import it directly like import { lateralDrawer as lateralDrawerStyles } from "@tasty-maker-studio/styled-system/recipes"; then it will work as expected

Once I copy these src files changes ^^^^ it began working. 🥇 

You're  :cherries:  awesome Thanks for sticking with me..
